src/libs/sys-util-rs/src/mount.rs

```diff
@@ -0,0 +1,808 @@
+}
+
+pub use kata_types::k8s::is_k8s_empty_dir;
+
+/// Check whether the given path is a kubernetes ephemeral volume.
+///
+/// This method depends on a specific path used by k8s to detect if it's type of ephemeral.
+/// As of now, this is a very k8s specific solution that works but in future there should be a
+/// better way for this method to determine if the path is for ephemeral volume type.
+pub fn is_k8s_ephemeral_volume(path: &str) -> bool {
+    if is_k8s_empty_dir(path) {
+        if let Ok((_dev_path, fs_type)) = get_device_path_and_fs_type(path) {
+            if fs_type == "tmpfs" {
+                return true;
+            }
+        }
+    }
+
+    false
+}
+
+/// Check whether the given path is a kubernetes empty-dir volume of medium "default".
+///
+/// K8s `EmptyDir` volumes are directories on the host. If the fs type is tmpfs, it's a ephemeral
+/// volume instead of a `EmptyDir` volume.
+pub fn is_k8s_host_empty_dir(path: &str) -> bool {
+    if is_k8s_empty_dir(path) {
+        if let Ok((_dev_path, fs_type)) = get_device_path_and_fs_type(path) {
+            if fs_type != "tmpfs" {
+                return true;
+            }
+        }
+    }
+
+    false
+}
+
+/// Remount a bind mountpoint in readonly mode.
+pub fn bind_remount_read_only<P: AsRef<Path>>(dst: P) -> Result<()> {
+    let dst = dst.as_ref();
+    if dst.as_os_str().is_empty() {
+        return Err(Error::EmptyMountPoint);
+    }
+    let dst = dst
+        .canonicalize()
+        .map_err(|_e| Error::InvalidPath(dst.to_string_lossy().to_string()))?;
+
+    mount(
+        Some(""),
+        &dst,
+        Some("bind"),
+        MsFlags::MS_BIND | MsFlags::MS_REMOUNT | MsFlags::MS_RDONLY,
+        Some(""),
+    )
+    .map_err(|e| Error::BindMount("".to_string(), dst.to_string_lossy().to_string(), e))
+}
+
+/// Bind mount `src` to `dst` in slave mode.
+pub fn bind_mount<P: AsRef<Path>>(src: P, dst: P, read_only: bool) -> Result<()> {
+    fail::fail_point!("bind_mount", |_| {
+        Err(anyhow!("Bind mount fail point injection"))
+    });
+
+    let src = src.as_ref();
+    let dst = dst.as_ref();
+    if src.as_os_str().is_empty() {
+        return Err(Error::EmptyMountPoint);
+    }
+    if dst.as_os_str().is_empty() {
+        return Err(Error::EmptyMountPoint);
+    }
+    let abs_src = src
+        .canonicalize()
+        .map_err(|_e| Error::InvalidPath(src.to_string_lossy().to_string()))?;
+
+    ensure_dst_exists(abs_src.as_path(), dst)?;
+    mount(
+        Some(&abs_src),
+        dst,
+        Some("bind"),
+        MsFlags::MS_BIND,
+        Some(""),
+    )
+    .map_err(|e| {
+        Error::BindMount(
+            abs_src.to_string_lossy().to_string(),
+            dst.to_string_lossy().to_string(),
+            e,
+        )
+    })?;
+
+    mount(Some("none"), dst, Some(""), MsFlags::MS_SLAVE, Some(""))
+        .map_err(|e| Error::Mount("".to_string(), dst.to_string_lossy().to_string(), e))?;
+
+    if read_only {
```

Comment:
It's a limitation of linux mount syscall:
```
   Creating a bind mount
       If mountflags includes MS_BIND (available since Linux 2.4), then
       perform a bind mount.  A bind mount makes a file or a directory
       subtree visible at another point within the single directory
       hierarchy.  Bind mounts may cross filesystem boundaries and span
       chroot(2) jails.

       The filesystemtype and data arguments are ignored.

       The remaining bits (other than MS_REC, described below) in the
       mountflags argument are also ignored.  (The bind mount has the
       same mount options as the underlying mount.)  However, see the
       discussion of remounting above, for a method of making an
       existing bind mount read-only.

       Since Linux 2.6.26, the MS_REMOUNT flag can be used with MS_BIND
       to modify only the per-mount-point flags.  This is particularly
       useful for setting or clearing the "read-only" flag on a mount
       without changing the underlying filesystem.  Specifying
       mountflags as:

           MS_REMOUNT | MS_BIND | MS_RDONLY

       will make access through this mountpoint read-only, without
       affecting other mounts.
```

Reply:
> @jiangliu - can you add a comment about that in the code for clarity?

I agree.